File: bot/src/telemetry.py
"""
Telemetry client — bot → API bridge.

Every significant event the bot produces is POSTed to the Node.js API,
which persists it to PostgreSQL and broadcasts it to dashboard WebSocket clients.

All calls are fire-and-forget (background thread) so the trading loop
is never blocked by network latency.

Usage:
    from telemetry import Telemetry
    tel = Telemetry()
    tel.session_start(equity=50000.0)
    tel.trade_entry(symbol="NVAX", ...)
"""

import logging
import os
import threading
import time
from datetime import date, datetime, timezone
from typing import Any, Optional

import requests

logger = logging.getLogger(__name__)

API_BASE = os.environ.get("API_BASE_URL", "http://api:4000")
TIMEOUT = 5  # seconds — never block the trading loop longer than this


class Telemetry:
    def __init__(self):
        self._session_date = str(date.today())
        self._enabled = bool(API_BASE)
        # Outstanding fire-and-forget threads — flush() waits for these.
        # The bot itself never calls flush (it runs forever, threads complete
        # naturally), but standalone scripts (test_trade, multi_test_trade)
        # need it so their final trade_exit POSTs land before exit.
        self._pending: list[threading.Thread] = []
        logger.info("Telemetry target: %s", API_BASE)

    # ------------------------------------------------------------------
    # Session lifecycle
    # ------------------------------------------------------------------

    def session_start(self, equity: float, trading_mode: str = "paper") -> None:
        self._session_date = str(date.today())
        # Synchronous so the session row exists before any subsequent event/scan
        # is posted — otherwise the API can't link them back to the session.
        self._post_sync("/telemetry/session/start", {
            "date": self._session_date,
            "startingEquity": equity,
            "tradingMode": trading_mode,
        })
        self.event("SESSION_START", f"Session started. Equity: ${equity:.2f}")

    def session_end(self, summary: dict) -> None:
        payload = {
            "date": self._session_date,
            "endingEquity": summary.get("ending_equity"),
            "realizedPnl": summary.get("pnl", 0),
            "totalTrades": summary.get("trades", 0),
            "winningTrades": summary.get("winners", 0),
            "losingTrades": summary.get("losers", 0),
            "accuracyPct": summary.get("accuracy_pct"),
            "avgWinner": summary.get("avg_winner"),
            "avgLoser": summary.get("avg_loser"),
            "halted": summary.get("halted", False),
            "haltReason": summary.get("halt_reason"),
        }
        self._post("/telemetry/session/end", payload)
        self.event("SESSION_END", f"Session ended. Day P&L: ${summary.get('pnl', 0):+.2f}")

    # ------------------------------------------------------------------
    # Trade lifecycle
    # ------------------------------------------------------------------

    def trade_entry(
        self,
        symbol: str,
        setup: str,
        qty: int,
        entry_price: float,
        stop_price: float,
        target_price: float,
        order_id: Optional[str] = None,
        vwap: Optional[float] = None,
        macd: Optional[float] = None,
        rvol: Optional[float] = None,
        signal_id: Optional[str] = None,
    ) -> Optional[str]:
        """Posts trade entry. Returns trade DB id (or None on failure)."""
        payload = {
            "sessionDate": self._session_date,
            "symbol": symbol,
            "setup": setup,
            "qty": qty,
            "entryPrice": entry_price,
            "stopPrice": stop_price,
            "targetPrice": target_price,
            "entryOrderId": order_id,
            "entryVwap": vwap,
            "entryMacd": macd,
            "entryRvol": rvol,
            "signalId": signal_id,
        }
        result = self._post_sync("/telemetry/trade/entry", payload)
        return result.get("id") if result else None

    def trade_exit(
        self,
        trade_id: str,
        exit_price: float,
        exit_reason: str,
        order_id: Optional[str] = None,
    ) -> None:
        self._post("/telemetry/trade/exit", {
            "tradeId": trade_id,
            "exitPrice": exit_price,
            "exitReason": exit_reason,
            "exitOrderId": order_id,
        })

    # ------------------------------------------------------------------
    # Signal evaluation
    # ------------------------------------------------------------------

    def signal(
        self,
        symbol: str,
        setup: str,
        confidence: str,
        entry_price: float,
        stop_price: float,
        target_price: float,
        acted: bool,
        rejection_reason: Optional[str] = None,
        vwap: Optional[float] = None,
        macd_line: Optional[float] = None,
        macd_histogram: Optional[float] = None,
        rvol: Optional[float] = None,
        price: Optional[float] = None,
        pct_change: Optional[float] = None,
    ) -> Optional[str]:
        """Posts signal. Returns signal DB id (or None on failure)."""
        payload = {
            "sessionDate": self._session_date,
            "symbol": symbol,
            "setup": setup,
            "confidence": confidence,
            "entryPrice": entry_price,
            "stopPrice": stop_price,
            "targetPrice": target_price,
            "acted": acted,
            "rejectionReason": rejection_reason,
            "vwap": vwap,
            "macdLine": macd_line,
            "macdHistogram": macd_histogram,
            "rvolAtSignal": rvol,
            "price": price,
            "pctChange": pct_change,
        }
        result = self._post_sync("/telemetry/signal", payload)
        return result.get("id") if result else None

    # ------------------------------------------------------------------
    # Scanner
    # ------------------------------------------------------------------

    def scan_result(self, candidates: list[dict], stats: Optional[dict] = None) -> None:
        self._post("/telemetry/scan", {
            "sessionDate": self._session_date,
            "candidates": candidates,
            "stats": stats or {},
        })

    # ------------------------------------------------------------------
    # Equity + positions
    # ------------------------------------------------------------------

    def equity_snapshot(
        self,
        equity: float,
        buying_power: Optional[float] = None,
        day_pnl: Optional[float] = None,
        open_position_count: int = 0,
    ) -> None:
        self._post("/telemetry/equity", {
            "sessionDate": self._session_date,
            "equity": equity,
            "buyingPower": buying_power,
            "dayPnl": day_pnl,
            "openPositionCount": open_position_count,
        })

    def position_update(self, symbol: str, current_price: float, unrealized_pnl: float) -> None:
        self._post("/telemetry/position", {
            "symbol": symbol,
            "currentPrice": current_price,
            "unrealizedPnl": unrealized_pnl,
        })

    def gate_check(
        self,
        symbol: str,
        gates: dict,
        setup: Optional[str] = None,
        confidence: Optional[str] = None,
    ) -> None:
        """Per-bar strategy gate evaluation. Fires on every evaluate() call."""
        self._post("/telemetry/gate-check", {
            "symbol": symbol,
            "gates": gates,
            "setup": setup,
            "confidence": confidence,
            "ts": datetime.now(timezone.utc).isoformat(),
        })

    # ------------------------------------------------------------------
    # Generic events
    # ------------------------------------------------------------------

    def event(
        self,
        event_type: str,
        message: str,
        severity: str = "INFO",
        metadata: Optional[dict] = None,
    ) -> None:
        self._post("/telemetry/event", {
            "sessionDate": self._session_date,
            "eventType": event_type,
            "severity": severity,
            "message": message,
            "metadata": metadata,
        })

    def error(self, message: str, metadata: Optional[dict] = None) -> None:
        self.event("INFO", message, severity="ERROR", metadata=metadata)
        logger.error("Telemetry ERROR: %s", message)

    # ------------------------------------------------------------------
    # HTTP helpers
    # ------------------------------------------------------------------

    def _post(self, path: str, payload: dict) -> None:
        """Fire-and-forget in background thread."""
        if not self._enabled:
            return
        t = threading.Thread(
            target=self._post_sync, args=(path, payload), daemon=True
        )
        t.start()
        # Track for flush(); cap the list so a long-running bot doesn't grow
        # this unbounded. We only really need recent threads for flush at exit.
        self._pending.append(t)
        if len(self._pending) > 200:
            self._pending = [x for x in self._pending if x.is_alive()][-100:]

    def flush(self, timeout: float = 10.0) -> None:
        """Wait up to `timeout` seconds for outstanding fire-and-forget POSTs
        to complete. Call this from a standalone script before exiting so
        the final trade_exit / event posts have a chance to land."""
        deadline = time.time() + timeout
        for t in list(self._pending):
            remaining = deadline - time.time()
            if remaining <= 0:
                break
            t.join(timeout=remaining)
        self._pending = [t for t in self._pending if t.is_alive()]

    def _post_sync(self, path: str, payload: dict) -> dict:
        """Blocking POST — use only when return value is needed."""
        if not self._enabled:
            return {}
        try:
            resp = requests.post(
                f"{API_BASE}{path}",
                json=payload,
                timeout=TIMEOUT,
            )
            if resp.status_code >= 400:
                logger.warning("Telemetry POST %s → %d: %s", path, resp.status_code, resp.text[:200])
                return {}
            return resp.json()
        except Exception as e:
            logger.debug("Telemetry POST failed (%s): %s", path, e)
            return {}

File: api/src/routes/telemetry.ts
/**
 * Telemetry ingestion endpoint — called by the Python bot.
 *
 * The bot POSTs structured events here. Each event is persisted to
 * PostgreSQL and broadcast to all connected WebSocket clients in real time.
 *
 * POST /telemetry/event        — generic bot event
 * POST /telemetry/trade/entry  — new trade opened
 * POST /telemetry/trade/exit   — trade closed
 * POST /telemetry/signal       — signal evaluated (acted on or not)
 * POST /telemetry/scan         — scanner result
 * POST /telemetry/equity       — equity snapshot
 * POST /telemetry/position     — open position update
 * POST /telemetry/session/start
 * POST /telemetry/session/end
 */

import { Router, Request, Response } from "express";
import { prisma } from "../index";
import { broadcast } from "../websocket";

const router = Router();

// Per-symbol latest gate-check, kept in process memory only. Cheap and high
// frequency (one per bar per watchlist symbol), so persisting would just
// bloat Postgres. Cleared every 30 min so stale entries don't pile up.
interface GateCheckRecord {
  symbol: string;
  gates: Record<string, boolean>;
  setup: string | null;
  confidence: string | null;
  ts: string;
  receivedAt: number;
}
const gateCache = new Map<string, GateCheckRecord>();
setInterval(() => {
  const cutoff = Date.now() - 30 * 60 * 1000;
  for (const [sym, rec] of gateCache.entries()) {
    if (rec.receivedAt < cutoff) gateCache.delete(sym);
  }
}, 60 * 1000);
export function getGateCache(): GateCheckRecord[] {
  return [...gateCache.values()].sort((a, b) => b.receivedAt - a.receivedAt);
}

// ----------------------------------------------------------------
// Session start
// ----------------------------------------------------------------
router.post("/session/start", async (req: Request, res: Response) => {
  const { date, startingEquity, tradingMode } = req.body;
  try {
    const session = await prisma.tradingSession.upsert({
      where: { date: new Date(date) },
      update: { startedAt: new Date(), startingEquity, tradingMode },
      create: { date: new Date(date), startingEquity, tradingMode },
    });
    broadcast(req.app.locals.wss, "session_start", session);
    res.json(session);
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

// ----------------------------------------------------------------
// Session end
// ----------------------------------------------------------------
router.post("/session/end", async (req: Request, res: Response) => {
  const { date, endingEquity, realizedPnl, totalTrades, winningTrades,
          losingTrades, accuracyPct, avgWinner, avgLoser, halted, haltReason } = req.body;
  try {
    const session = await prisma.tradingSession.update({
      where: { date: new Date(date) },
      data: { endedAt: new Date(), endingEquity, realizedPnl, totalTrades,
               winningTrades, losingTrades, accuracyPct, avgWinner, avgLoser,
               halted, haltReason },
    });
    broadcast(req.app.locals.wss, "session_end", session);
    res.json(session);
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

// ----------------------------------------------------------------
// Trade entry
// ----------------------------------------------------------------
router.post("/trade/entry", async (req: Request, res: Response) => {
  const { sessionDate, symbol, setup, qty, entryPrice, stopPrice, targetPrice,
          entryOrderId, entryVwap, entryMacd, entryRvol, signalId } = req.body;
  try {
    const session = await prisma.tradingSession.findUnique({
      where: { date: new Date(sessionDate) },
    });
    if (!session) return res.status(404).json({ error: "Session not found" });

    const riskPerShare = entryPrice - stopPrice;
    const rewardPerShare = targetPrice - entryPrice;
    const trade = await prisma.trade.create({
      data: {
        sessionId: session.id,
        symbol, setup, qty,
        entryPrice, stopPrice, targetPrice,
        riskPerShare, rewardPerShare,
        rrRatio: riskPerShare > 0 ? rewardPerShare / riskPerShare : 0,
        entryTime: new Date(),
        entryOrderId, entryVwap, entryMacd, entryRvol,
        signalId: signalId || null,
        status: "OPEN",
      },
    });

    // Update open position record
    await prisma.position.upsert({
      where: { symbol },
      update: { qty, entryPrice, stopPrice, targetPrice, setup, entryTime: new Date() },
      create: { symbol, qty, entryPrice, stopPrice, targetPrice, setup, entryTime: new Date() },
    });

    broadcast(req.app.locals.wss, "trade_entry", trade);
    res.json(trade);
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

// ----------------------------------------------------------------
// Trade exit
// ----------------------------------------------------------------
router.post("/trade/exit", async (req: Request, res: Response) => {
  const { tradeId, exitPrice, exitReason, exitOrderId } = req.body;
  try {
    const existing = await prisma.trade.findUnique({ where: { id: tradeId } });
    if (!existing) return res.status(404).json({ error: "Trade not found" });

    const realizedPnl = (exitPrice - existing.entryPrice) * existing.qty;
    const pnlPct = ((exitPrice - existing.entryPrice) / existing.entryPrice) * 100;
    const holdMs = Date.now() - existing.entryTime.getTime();
    const holdMinutes = Math.round(holdMs / 60000);

    const trade = await prisma.trade.update({
      where: { id: tradeId },
      data: { exitPrice, exitTime: new Date(), exitReason, exitOrderId,
               realizedPnl, pnlPct, holdMinutes, status: "CLOSED" },
    });

    // Remove from open positions
    await prisma.position.deleteMany({ where: { symbol: existing.symbol } });

    broadcast(req.app.locals.wss, "trade_exit", trade);
    res.json(trade);
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

// ----------------------------------------------------------------
// Signal evaluated
// ----------------------------------------------------------------
router.post("/signal", async (req: Request, res: Response) => {
  const { sessionDate, symbol, setup, confidence, entryPrice, stopPrice,
          targetPrice, acted, rejectionReason, vwap, macdLine, macdHistogram,
          rvolAtSignal, price, pctChange } = req.body;
  try {
    const session = await prisma.tradingSession.findUnique({
      where: { date: new Date(sessionDate) },
    });
    if (!session) return res.status(404).json({ error: "Session not found" });

    const rrRatio = stopPrice < entryPrice
      ? (targetPrice - entryPrice) / (entryPrice - stopPrice) : 0;

    const signal = await prisma.signal.create({
      data: {
        sessionId: session.id,
        symbol, setup, confidence,
        entryPrice, stopPrice, targetPrice, rrRatio,
        acted, rejectionReason,
        vwap, macdLine, macdHistogram, rvolAtSignal, price, pctChange,
      },
    });
    broadcast(req.app.locals.wss, "signal", signal);
    res.json(signal);
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

// ----------------------------------------------------------------
// Scanner result
// ----------------------------------------------------------------
router.post("/scan", async (req: Request, res: Response) => {
  const { sessionDate, candidates, stats } = req.body;
  try {
    const session = await prisma.tradingSession.findUnique({
      where: { date: new Date(sessionDate) },
    });
    if (!session) return res.status(404).json({ error: "Session not found" });

    const passed = candidates.filter((c: any) => c.passedFilters);
    const s = stats ?? {};
    const scanResult = await prisma.scanResult.create({
      data: {
        sessionId: session.id,
        candidatesFound: passed.length,
        universeSize:  s.universeSize  ?? null,
        evaluated:     s.evaluated     ?? null,
        rejectedPrice: s.rejectedPrice ?? null,
        rejectedPct:   s.rejectedPct   ?? null,
        rejectedRvol:  s.rejectedRvol  ?? null,
        rejectedFloat: s.rejectedFloat ?? null,
        durationMs:    s.durationMs    ?? null,
        candidates: {
          create: candidates.map((c: any, i: number) => ({ ...c, rank: i + 1 })),
        },
      },
      include: { candidates: true },
    });

    // Upsert WatchlistEntry for each candidate that passed filters
    for (const c of passed) {
      await prisma.watchlistEntry.create({
        data: {
          sessionId: session.id,
          symbol: c.symbol,
          addReason: "periodic_scan",
          score: c.score ?? null,
        },
      });
    }

    broadcast(req.app.locals.wss, "scan_complete", scanResult);
    res.json(scanResult);
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

// ----------------------------------------------------------------
// Equity snapshot
// ----------------------------------------------------------------
router.post("/equity", async (req: Request, res: Response) => {
  const { sessionDate, equity, buyingPower, dayPnl, openPositionCount } = req.body;
  try {
    const session = await prisma.tradingSession.findFirst({
      where: { date: new Date(sessionDate) },
    });
    const snap = await prisma.equitySnapshot.create({
      data: { sessionId: session?.id, equity, buyingPower, dayPnl, openPositionCount },
    });
    broadcast(req.app.locals.wss, "equity_update", snap);
    res.json(snap);
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

// ----------------------------------------------------------------
// Position update (live unrealized PnL)
// ----------------------------------------------------------------
router.post("/position", async (req: Request, res: Response) => {
  const { symbol, currentPrice, unrealizedPnl } = req.body;
  try {
    const pos = await prisma.position.updateMany({
      where: { symbol },
      data: { currentPrice, unrealizedPnl },
    });
    broadcast(req.app.locals.wss, "position_update", { symbol, currentPrice, unrealizedPnl });
    res.json({ updated: pos.count });
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

// ----------------------------------------------------------------
// Per-bar gate-check from the strategy. Stored in-memory only and
// broadcast over WebSocket so the dashboard's TradeEntryGates panel can
// flip per-symbol pulses live.
// ----------------------------------------------------------------
router.post("/gate-check", async (req: Request, res: Response) => {
  const { symbol, gates, setup, confidence, ts } = req.body;
  if (!symbol || !gates) return res.status(400).json({ error: "symbol and gates required" });
  const record: GateCheckRecord = {
    symbol,
    gates,
    setup: setup ?? null,
    confidence: confidence ?? null,
    ts: ts ?? new Date().toISOString(),
    receivedAt: Date.now(),
  };
  gateCache.set(symbol, record);
  broadcast(req.app.locals.wss, "gate_check", record);
  res.json({ ok: true });
});

// ----------------------------------------------------------------
// Generic bot event
// ----------------------------------------------------------------
router.post("/event", async (req: Request, res: Response) => {
  const { sessionDate, eventType, severity, message, metadata } = req.body;
  try {
    const session = sessionDate
      ? await prisma.tradingSession.findFirst({ where: { date: new Date(sessionDate) } })
      : null;
    const event = await prisma.botEvent.create({
      data: { sessionId: session?.id, eventType, severity: severity || "INFO", message, metadata },
    });
    broadcast(req.app.locals.wss, "bot_event", event);
    res.json(event);
  } catch (e: any) {
    res.status(500).json({ error: e.message });
  }
});

export default router;
